In [1]:
# Imports
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import sklearn

In [2]:
pd.show_versions()


INSTALLED VERSIONS
------------------
commit              : a60ad39b4a9febdea9a59d602dad44b1538b0ea5
python              : 3.10.12.final.0
python-bits         : 64
OS                  : Linux
OS-release          : 6.2.0-37-generic
Version             : #38~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Nov  2 18:01:13 UTC 2
machine             : x86_64
processor           : x86_64
byteorder           : little
LC_ALL              : None
LANG                : fr_FR.UTF-8
LOCALE              : fr_FR.UTF-8

pandas              : 2.1.2
numpy               : 1.26.1
pytz                : 2022.1
dateutil            : 2.8.2
setuptools          : 59.6.0
pip                 : 22.0.2
Cython              : 3.0.4
pytest              : None
hypothesis          : None
sphinx              : None
blosc               : None
feather             : None
xlsxwriter          : None
lxml.etree          : None
html5lib            : None
pymysql             : None
psycopg2            : None
jinja2              : 3.1.2


In [3]:
pd.set_option('display.max_columns',50)
pd.set_option('display.max_rows',500)

In [4]:
# Chargement des données
app_test = pd.read_csv("application_test.csv")
app_train = pd.read_csv("application_train.csv")
bureau_balance = pd.read_csv("bureau_balance.csv")
bureau = pd.read_csv("bureau.csv")
credit_card_balance = pd.read_csv("credit_card_balance.csv")
HomeCredit_columns_description = pd.read_csv("HomeCredit_columns_description.csv",encoding='latin-1')
installments_payments = pd.read_csv("installments_payments.csv")
POS_CASH_balance = pd.read_csv("POS_CASH_balance.csv")
previous_app = pd.read_csv("previous_application.csv")
sample_submission = pd.read_csv("sample_submission.csv")


# Analyse exploratoire des données

In [5]:
# Affichage des dataframes
display(app_test.head())
display(app_train.head())
display(bureau_balance.head())
display(bureau.head())
display(credit_card_balance.head())
display(HomeCredit_columns_description.head())
display(installments_payments.head())
display(POS_CASH_balance.head())
display(previous_app.head())
display(sample_submission.head())

SK_ID_CURR NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  \
0      100001         Cash loans           F            N               Y   
1      100005         Cash loans           M            N               Y   
2      100013         Cash loans           M            Y               Y   
3      100028         Cash loans           F            N               Y   
4      100038         Cash loans           M            Y               N   

   CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  AMT_GOODS_PRICE  \
0             0          135000.0    568800.0      20560.5         450000.0   
1             0           99000.0    222768.0      17370.0         180000.0   
2             0          202500.0    663264.0      69777.0         630000.0   
3             2          315000.0   1575000.0      49018.5        1575000.0   
4             1          180000.0    625500.0      32067.0         625500.0   

  NAME_TYPE_SUITE NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0   Unaccompanied          Working               Higher education   
1   Unaccompanied          Working  Secondary / secondary special   
2             NaN          Working               Higher education   
3   Unaccompanied          Working  Secondary / secondary special   
4   Unaccompanied          Working  Secondary / secondary special   

  NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  REGION_POPULATION_RELATIVE  \
0            Married  House / apartment                    0.018850   
1            Married  House / apartment                    0.035792   
2            Married  House / apartment                    0.019101   
3            Married  House / apartment                    0.026392   
4            Married  House / apartment                    0.010032   

   DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  DAYS_ID_PUBLISH  OWN_CAR_AGE  \
0      -19241          -2329            -5170.0             -812          NaN   
1      -18064          -4469            -9118.0            -1623          NaN   
2      -20038          -4458            -2175.0            -3503          5.0   
3      -13976          -1866            -2000.0            -4208          NaN   
4      -13040          -2191            -4000.0            -4262         16.0   

   FLAG_MOBIL  FLAG_EMP_PHONE  FLAG_WORK_PHONE  FLAG_CONT_MOBILE  ...  \
0           1               1                0                 1  ...   
1           1               1                0                 1  ...   
2           1               1                0                 1  ...   
3           1               1                0                 1  ...   
4           1               1                1                 1  ...   

   FLAG_DOCUMENT_3  FLAG_DOCUMENT_4 FLAG_DOCUMENT_5  FLAG_DOCUMENT_6  \
0                1                0               0                0   
1                1                0               0                0   
2                0                0               0                0   
3                1                0               0                0   
4                1                0               0                0   

   FLAG_DOCUMENT_7  FLAG_DOCUMENT_8 FLAG_DOCUMENT_9  FLAG_DOCUMENT_10  \
0                0                0               0                 0   
1                0                0               0                 0   
2                0                1               0                 0   
3                0                0               0                 0   
4                0                0               0                 0   

   FLAG_DOCUMENT_11  FLAG_DOCUMENT_12  FLAG_DOCUMENT_13  FLAG_DOCUMENT_14  \
0                 0                 0                 0                 0   
1                 0                 0                 0                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0   

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   

   AMT_GOODS_PRICE NAME_TYPE_SUITE NAME_INCOME_TYPE  \
0         351000.0   Unaccompanied          Working   
1        1129500.0          Family    State servant   
2         135000.0   Unaccompanied          Working   
3         297000.0   Unaccompanied          Working   
4         513000.0   Unaccompanied          Working   

             NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  \
0  Secondary / secondary special  Single / not married  House / apartment   
1               Higher education               Married  House / apartment   
2  Secondary / secondary special  Single / not married  House / apartment   
3  Secondary / secondary special        Civil marriage  House / apartment   
4  Secondary / secondary special  Single / not married  House / apartment   

   REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  \
0                    0.018801       -9461           -637            -3648.0   
1                    0.003541      -16765          -1188            -1186.0   
2                    0.010032      -19046           -225            -4260.0   
3                    0.008019      -19005          -3039            -9833.0   
4                    0.028663      -19932          -3038            -4311.0   

   DAYS_ID_PUBLISH  OWN_CAR_AGE  FLAG_MOBIL  FLAG_EMP_PHONE  FLAG_WORK_PHONE  \
0            -2120          NaN           1               1                0   
1             -291          NaN           1               1                0   
2            -2531         26.0           1               1                1   
3            -2437          NaN           1               1                0   
4            -3458          NaN           1               1                0   

   ...  FLAG_DOCUMENT_3  FLAG_DOCUMENT_4  FLAG_DOCUMENT_5 FLAG_DOCUMENT_6  \
0  ...                1                0                0               0   
1  ...                1                0                0               0   
2  ...                0                0                0               0   
3  ...                1                0                0               0   
4  ...                0                0                0               0   

   FLAG_DOCUMENT_7  FLAG_DOCUMENT_8  FLAG_DOCUMENT_9 FLAG_DOCUMENT_10  \
0                0                0                0                0   
1                0                0                0                0   
2                0                0                0                0   
3                0                0                0                0   
4                0                1                0                0   

   FLAG_DOCUMENT_11  FLAG_DOCUMENT_12  FLAG_DOCUMENT_13  FLAG_DOCUMENT_14  \
0                 0                 0                 0                 0   
1                 0                 0                 0                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 0                 0   

   F

SK_ID_BUREAU  MONTHS_BALANCE STATUS
0       5715448               0      C
1       5715448              -1      C
2       5715448              -2      C
3       5715448              -3      C
4       5715448              -4      C

SK_ID_CURR  SK_ID_BUREAU CREDIT_ACTIVE CREDIT_CURRENCY  DAYS_CREDIT  \
0      215354       5714462        Closed      currency 1         -497   
1      215354       5714463        Active      currency 1         -208   
2      215354       5714464        Active      currency 1         -203   
3      215354       5714465        Active      currency 1         -203   
4      215354       5714466        Active      currency 1         -629   

   CREDIT_DAY_OVERDUE  DAYS_CREDIT_ENDDATE  DAYS_ENDDATE_FACT  \
0                   0               -153.0             -153.0   
1                   0               1075.0                NaN   
2                   0                528.0                NaN   
3                   0                  NaN                NaN   
4                   0               1197.0                NaN   

   AMT_CREDIT_MAX_OVERDUE  CNT_CREDIT_PROLONG  AMT_CREDIT_SUM  \
0                     NaN                   0         91323.0   
1                     NaN                   0        225000.0   
2                     NaN                   0        464323.5   
3                     NaN                   0         90000.0   
4                 77674.5                   0       2700000.0   

   AMT_CREDIT_SUM_DEBT  AMT_CREDIT_SUM_LIMIT  AMT_CREDIT_SUM_OVERDUE  \
0                  0.0                   NaN                     0.0   
1             171342.0                   NaN                     0.0   
2                  NaN                   NaN                     0.0   
3                  NaN                   NaN                     0.0   
4                  NaN                   NaN                     0.0   

       CREDIT_TYPE  DAYS_CREDIT_UPDATE  AMT_ANNUITY  
0  Consumer credit                -131          NaN  
1      Credit card                 -20          NaN  
2  Consumer credit                 -16          NaN  
3      Credit card                 -16          NaN  
4  Consumer credit                 -21          NaN

SK_ID_PREV  SK_ID_CURR  MONTHS_BALANCE  AMT_BALANCE  \
0     2562384      378907              -6       56.970   
1     2582071      363914              -1    63975.555   
2     1740877      371185              -7    31815.225   
3     1389973      337855              -4   236572.110   
4     1891521      126868              -1   453919.455   

   AMT_CREDIT_LIMIT_ACTUAL  AMT_DRAWINGS_ATM_CURRENT  AMT_DRAWINGS_CURRENT  \
0                   135000                       0.0                 877.5   
1                    45000                    2250.0                2250.0   
2                   450000                       0.0                   0.0   
3                   225000                    2250.0                2250.0   
4                   450000                       0.0               11547.0   

   AMT_DRAWINGS_OTHER_CURRENT  AMT_DRAWINGS_POS_CURRENT  \
0                         0.0                     877.5   
1                         0.0                       0.0   
2                         0.0                       0.0   
3                         0.0                       0.0   
4                         0.0                   11547.0   

   AMT_INST_MIN_REGULARITY  AMT_PAYMENT_CURRENT  AMT_PAYMENT_TOTAL_CURRENT  \
0                 1700.325               1800.0                     1800.0   
1                 2250.000               2250.0                     2250.0   
2                 2250.000               2250.0                     2250.0   
3                11795.760              11925.0                    11925.0   
4                22924.890              27000.0                    27000.0   

   AMT_RECEIVABLE_PRINCIPAL  AMT_RECIVABLE  AMT_TOTAL_RECEIVABLE  \
0                     0.000          0.000                 0.000   
1                 60175.080      64875.555             64875.555   
2                 26926.425      31460.085             31460.085   
3                224949.285     233048.970            233048.970   
4                443044.395     453919.455            453919.455   

   CNT_DRAWINGS_ATM_CURRENT  CNT_DRAWINGS_CURRENT  CNT_DRAWINGS_OTHER_CURRENT  \
0                       0.0                     1                         0.0   
1                       1.0                     1                         0.0   
2                       0.0                     0                         0.0   
3                       1.0                     1                         0.0   
4                       0.0                     1                         0.0   

   CNT_DRAWINGS_POS_CURRENT  CNT_INSTALMENT_MATURE_CUM NAME_CONTRACT_STATUS  \
0                       1.0                       35.0               Active   
1                       0.0                       69.0               Active   
2                       0.0                       30.0               Active   
3                       0.0                       10.0               Active   
4                       1.0                      101.0               Active   

   SK_DPD  SK_DPD_DEF  
0       0           0  
1       0           0  
2       0           0  
3       0           0  
4       0           0

Unnamed: 0                         Table                 Row  \
0           1  application_{train|test}.csv          SK_ID_CURR   
1           2  application_{train|test}.csv              TARGET   
2           5  application_{train|test}.csv  NAME_CONTRACT_TYPE   
3           6  application_{train|test}.csv         CODE_GENDER   
4           7  application_{train|test}.csv        FLAG_OWN_CAR   

                                         Description Special  
0                           ID of loan in our sample     NaN  
1  Target variable (1 - client with payment diffi...     NaN  
2        Identification if loan is cash or revolving     NaN  
3                               Gender of the client     NaN  
4                      Flag if the client owns a car     NaN

SK_ID_PREV  SK_ID_CURR  NUM_INSTALMENT_VERSION  NUM_INSTALMENT_NUMBER  \
0     1054186      161674                     1.0                      6   
1     1330831      151639                     0.0                     34   
2     2085231      193053                     2.0                      1   
3     2452527      199697                     1.0                      3   
4     2714724      167756                     1.0                      2   

   DAYS_INSTALMENT  DAYS_ENTRY_PAYMENT  AMT_INSTALMENT  AMT_PAYMENT  
0          -1180.0             -1187.0        6948.360     6948.360  
1          -2156.0             -2156.0        1716.525     1716.525  
2            -63.0               -63.0       25425.000    25425.000  
3          -2418.0             -2426.0       24350.130    24350.130  
4          -1383.0             -1366.0        2165.040     2160.585

SK_ID_PREV  SK_ID_CURR  MONTHS_BALANCE  CNT_INSTALMENT  \
0     1803195      182943             -31            48.0   
1     1715348      367990             -33            36.0   
2     1784872      397406             -32            12.0   
3     1903291      269225             -35            48.0   
4     2341044      334279             -35            36.0   

   CNT_INSTALMENT_FUTURE NAME_CONTRACT_STATUS  SK_DPD  SK_DPD_DEF  
0                   45.0               Active       0           0  
1                   35.0               Active       0           0  
2                    9.0               Active       0           0  
3                   42.0               Active       0           0  
4                   35.0               Active       0           0

SK_ID_PREV  SK_ID_CURR NAME_CONTRACT_TYPE  AMT_ANNUITY  AMT_APPLICATION  \
0     2030495      271877     Consumer loans     1730.430          17145.0   
1     2802425      108129         Cash loans    25188.615         607500.0   
2     2523466      122040         Cash loans    15060.735         112500.0   
3     2819243      176158         Cash loans    47041.335         450000.0   
4     1784265      202054         Cash loans    31924.395         337500.0   

   AMT_CREDIT  AMT_DOWN_PAYMENT  AMT_GOODS_PRICE WEEKDAY_APPR_PROCESS_START  \
0     17145.0               0.0          17145.0                   SATURDAY   
1    679671.0               NaN         607500.0                   THURSDAY   
2    136444.5               NaN         112500.0                    TUESDAY   
3    470790.0               NaN         450000.0                     MONDAY   
4    404055.0               NaN         337500.0                   THURSDAY   

   HOUR_APPR_PROCESS_START FLAG_LAST_APPL_PER_CONTRACT  \
0                       15                           Y   
1                       11                           Y   
2                       11                           Y   
3                        7                           Y   
4                        9                           Y   

   NFLAG_LAST_APPL_IN_DAY  RATE_DOWN_PAYMENT  RATE_INTEREST_PRIMARY  \
0                       1                0.0               0.182832   
1                       1                NaN                    NaN   
2                       1                NaN                    NaN   
3                       1                NaN                    NaN   
4                       1                NaN                    NaN   

   RATE_INTEREST_PRIVILEGED NAME_CASH_LOAN_PURPOSE NAME_CONTRACT_STATUS  \
0                  0.867336                    XAP             Approved   
1                       NaN                    XNA             Approved   
2                       NaN                    XNA             Approved   
3                       NaN                    XNA             Approved   
4                       NaN                Repairs              Refused   

   DAYS_DECISION      NAME_PAYMENT_TYPE CODE_REJECT_REASON  NAME_TYPE_SUITE  \
0            -73  Cash through the bank                XAP              NaN   
1           -164                    XNA                XAP    Unaccompanied   
2           -301  Cash through the bank                XAP  Spouse, partner   
3           -512  Cash through the bank                XAP              NaN   
4           -781  Cash through the bank                 HC              NaN   

  NAME_CLIENT_TYPE NAME_GOODS_CATEGORY NAME_PORTFOLIO NAME_PRODUCT_TYPE  \
0         Repeater              Mobile            POS               XNA   
1         Repeater                 XNA           Cash            x-sell   
2         Repeater                 XNA           Cash            x-sell   
3         Repeater                 XNA           Cash            x-sell   
4         Repeater                 XNA           Cash           walk-in   

              CHANNEL_TYPE  SELLERPLACE_AREA NAME_SELLER_INDUSTRY  \
0             Country-wide                35         Connectivity   
1           Contact center                -1                  XNA   
2  Credit and cash offices                -1                  XNA   
3  Credit and cash offices                -1                  XNA   
4  Credit and cash offices                -1                  XNA   

   CNT_PAYMENT NAME_YIELD_GROUP       PRODUCT_COMBINATION  DAYS_FIRST_DRAWING  \
0         12.0           middle  POS mobile with interest            365243.0   
1         36.0       low_action          Cash X-Sell: low            365243.0   
2         12.0             high         Cash X-Sell: high            365243.0   
3         12.0           middle       Cash X-Sell: middle            365243.0   
4         24.0             high         Cash Street: high                 NaN   

   D

SK_ID_CURR  TARGET
0      100001     0.5
1      100005     0.5
2      100013     0.5
3      100028     0.5
4      100038     0.5

In [6]:
## Fonctions pour descriptions des dataframes
list_dataset = []
list_shape = []
list_variable = []
list_duplicates = []
list_unique = []
list_null = []
list_null_ratio = []
list_type = []

dic_var_info = {'Dataset': list_dataset,
                'Shape': list_shape,
                'Variable': list_variable,
                'Duplicates': list_duplicates,
                'Unique': list_unique,
                'Null': list_null,
                'Null_ratio': list_null_ratio,
                'Type': list_type}


def var_info(df,df_name,var,print_=0,n=5):

    list_dataset.append(df_name)
    list_shape.append(df.shape)
    list_variable.append(var)
    list_duplicates.append(df.duplicated(subset=[var]).sum())
    list_unique.append(df[var].unique().shape[0])
    list_null.append(df[var].isna().sum().sum())
    list_null_ratio.append((df[var].isna().sum().sum())/df.shape[0])
    list_type.append(df[var].dtypes)

    df_data_model = pd.DataFrame(dic_var_info)

    if print_ == 1:
        varInfo = df_data_model.tail(n)
        return varInfo


def all_var_info(df,df_name):

    last_value = df.columns.tolist()[len(df.columns.tolist())-1]
    for i in df.columns.tolist():
        if i != last_value:
            var_info(df,df_name,i)
    return var_info(df,df_name,last_value,1,df.shape[1])

## Description des tables

### 1) app_train

 Il s'agit de la source de données principale.
 Chaque ligne représente une demande de crédit identifiée par la variable SK_ID_CURR (clé primaire).

Ce jeu de données comprend également la variable cible (target) qui peut prendre 2 valeurs (0 si le crédit a été remboursé et 1 si le crédit n'a pas été remboursé).

In [7]:
## Informations sur le premier dataframe: app_train
all_var_info(app_train,'app_train')

Dataset          Shape                      Variable  Duplicates  \
0    app_train  (307511, 122)                    SK_ID_CURR           0   
1    app_train  (307511, 122)                        TARGET      307509   
2    app_train  (307511, 122)            NAME_CONTRACT_TYPE      307509   
3    app_train  (307511, 122)                   CODE_GENDER      307508   
4    app_train  (307511, 122)                  FLAG_OWN_CAR      307509   
5    app_train  (307511, 122)               FLAG_OWN_REALTY      307509   
6    app_train  (307511, 122)                  CNT_CHILDREN      307496   
7    app_train  (307511, 122)              AMT_INCOME_TOTAL      304963   
8    app_train  (307511, 122)                    AMT_CREDIT      301908   
9    app_train  (307511, 122)                   AMT_ANNUITY      293838   
10   app_train  (307511, 122)               AMT_GOODS_PRICE      306508   
11   app_train  (307511, 122)               NAME_TYPE_SUITE      307503   
12   app_train  (307511, 122)              NAME_INCOME_TYPE      307503   
13   app_train  (307511, 122)           NAME_EDUCATION_TYPE      307506   
14   app_train  (307511, 122)            NAME_FAMILY_STATUS      307505   
15   app_train  (307511, 122)             NAME_HOUSING_TYPE      307505   
16   app_train  (307511, 122)    REGION_POPULATION_RELATIVE      307430   
17   app_train  (307511, 122)                    DAYS_BIRTH      290051   
18   app_train  (307511, 122)                 DAYS_EMPLOYED      294937   
19   app_train  (307511, 122)             DAYS_REGISTRATION      291823   
20   app_train  (307511, 122)               DAYS_ID_PUBLISH      301343   
21   app_train  (307511, 122)                   OWN_CAR_AGE      307448   
22   app_train  (307511, 122)                    FLAG_MOBIL      307509   
23   app_train  (307511, 122)                FLAG_EMP_PHONE      307509   
24   app_train  (307511, 122)               FLAG_WORK_PHONE      307509   
25   app_train  (307511, 122)              FLAG_CONT_MOBILE      307509   
26   app_train  (307511, 122)                    FLAG_PHONE      307509   
27   app_train  (307511, 122)                    FLAG_EMAIL      307509   
28   app_train  (307511, 122)               OCCUPATION_TYPE      307492   
29   app_train  (307511, 122)               CNT_FAM_MEMBERS      307493   
30   app_train  (307511, 122)          REGION_RATING_CLIENT      307508   
31   app_train  (307511, 122)   REGION_RATING_CLIENT_W_CITY      307508   
32   app_train  (307511, 122)    WEEKDAY_APPR_PROCESS_START      307504   
33   app_train  (307511, 122)       HOUR_APPR_PROCESS_START      307487   
34   app_train  (307511, 122)    REG_REGION_NOT_LIVE_REGION      307509   
35   app_train  (307511, 122)    REG_REGION_NOT_WORK_REGION      307509   
36   app_train  (307511, 122)   LIVE_REGION_NOT_WORK_REGION      307509   
37   app_train  (307511, 122)        REG_CITY_NOT_LIVE_CITY      307509   
38   app_train  (307511, 122)        REG_CITY_NOT_WORK_CITY      307509   
39   app_train  (307511, 122)       LIVE_CITY_NOT_WORK_CITY      307509   
40   app_train  (307511, 122)             ORGANIZATION_TYPE      307453   
41   app_train  (307511, 122)                  EXT_SOURCE_1      192926   
42   app_train  (307511, 122)                  EXT_SOURCE_2      187679   
43   app_train  (307511, 122)                  EXT_SOURCE_3      306696   
44   app_train  (307511, 122)                APARTMENTS_AVG      305171   
45   app_train  (307511, 122)              BASEMENTAREA_AVG      303730   
46   app_train  (307511, 122)   YEARS_BEGINEXPLUATATION_AVG      307225   
47   app_train  (307511, 122)               YEARS_BUILD_AVG      307361   
48   app_train  (307511, 122)                COMMONAREA_AVG      304329   
49   app_train  (307511, 122)                 ELEVATORS_AVG      307253   
50   app_train  (307511, 122)                 ENTRANCES_AVG      307225   
51   app_train  (307511, 122)                 FLOORSMAX_AVG      307107   
52   app_train  (307511, 122)   

### 2) app_test

Le jeu de données de test contient les mêmes variables que le jeu de données d'entraînement *excepté la variable cible*.

Nous utiliserons ce jeu de données pour réaliser des prédictions

In [8]:
## Informations sur le deuxième dataframe: app_test
all_var_info(app_test,'app_test')

Dataset         Shape                      Variable  Duplicates  Unique  \
122  app_test  (48744, 121)                    SK_ID_CURR           0   48744   
123  app_test  (48744, 121)            NAME_CONTRACT_TYPE       48742       2   
124  app_test  (48744, 121)                   CODE_GENDER       48742       2   
125  app_test  (48744, 121)                  FLAG_OWN_CAR       48742       2   
126  app_test  (48744, 121)               FLAG_OWN_REALTY       48742       2   
127  app_test  (48744, 121)                  CNT_CHILDREN       48733      11   
128  app_test  (48744, 121)              AMT_INCOME_TOTAL       48138     606   
129  app_test  (48744, 121)                    AMT_CREDIT       45807    2937   
130  app_test  (48744, 121)                   AMT_ANNUITY       41252    7492   
131  app_test  (48744, 121)               AMT_GOODS_PRICE       48067     677   
132  app_test  (48744, 121)               NAME_TYPE_SUITE       48736       8   
133  app_test  (48744, 121)              NAME_INCOME_TYPE       48737       7   
134  app_test  (48744, 121)           NAME_EDUCATION_TYPE       48739       5   
135  app_test  (48744, 121)            NAME_FAMILY_STATUS       48739       5   
136  app_test  (48744, 121)             NAME_HOUSING_TYPE       48738       6   
137  app_test  (48744, 121)    REGION_POPULATION_RELATIVE       48663      81   
138  app_test  (48744, 121)                    DAYS_BIRTH       33267   15477   
139  app_test  (48744, 121)                 DAYS_EMPLOYED       40881    7863   
140  app_test  (48744, 121)             DAYS_REGISTRATION       36126   12618   
141  app_test  (48744, 121)               DAYS_ID_PUBLISH       42864    5880   
142  app_test  (48744, 121)                   OWN_CAR_AGE       48691      53   
143  app_test  (48744, 121)                    FLAG_MOBIL       48742       2   
144  app_test  (48744, 121)                FLAG_EMP_PHONE       48742       2   
145  app_test  (48744, 121)               FLAG_WORK_PHONE       48742       2   
146  app_test  (48744, 121)              FLAG_CONT_MOBILE       48742       2   
147  app_test  (48744, 121)                    FLAG_PHONE       48742       2   
148  app_test  (48744, 121)                    FLAG_EMAIL       48742       2   
149  app_test  (48744, 121)               OCCUPATION_TYPE       48725      19   
150  app_test  (48744, 121)               CNT_FAM_MEMBERS       48732      12   
151  app_test  (48744, 121)          REGION_RATING_CLIENT       48741       3   
152  app_test  (48744, 121)   REGION_RATING_CLIENT_W_CITY       48740       4   
153  app_test  (48744, 121)    WEEKDAY_APPR_PROCESS_START       48737       7   
154  app_test  (48744, 121)       HOUR_APPR_PROCESS_START       48720      24   
155  app_test  (48744, 121)    REG_REGION_NOT_LIVE_REGION       48742       2   
156  app_test  (48744, 121)    REG_REGION_NOT_WORK_REGION       48742       2   
157  app_test  (48744, 121)   LIVE_REGION_NOT_WORK_REGION       48742       2   
158  app_test  (48744, 121)        REG_CITY_NOT_LIVE_CITY       48742       2   
159  app_test  (48744, 121)        REG_CITY_NOT_WORK_CITY       48742       2   
160  app_test  (48744, 121)       LIVE_CITY_NOT_WORK_CITY       48742       2   
161  app_test  (48744, 121)             ORGANIZATION_TYPE       48686      58   
162  app_test  (48744, 121)                  EXT_SOURCE_1       21536   27208   
163  app_test  (48744, 121)                  EXT_SOURCE_2        9858   38886   
164  app_test  (48744, 121)                  EXT_SOURCE_3       48041     703   
165  app_test  (48744, 121)                APARTMENTS_AVG       47200    1544   
166  app_test  (48744, 121)              BASEMENTAREA_AVG       45927    2817   
167  app_test  (48744, 121)   YEARS_BEGINEXPLUATATION_AVG       48568     176   
168  app_test  (48744, 121)               YEARS_BUILD_AVG       48613     131   
169  app_test  (48744, 121)                COMMONAREA_AVG       46701    2043   
170  app_test  (48744, 121)          

### 3) bureau

Données concernant les crédits antérieurs du client auprès d'autres institutions financières.
Chaque ligne concerne un crédit antérieur identifié par la variable SK_ID_BUREAU.

In [9]:
## Informations sur le troisième dataframe: bureau
all_var_info(bureau,'bureau')

Dataset          Shape                Variable  Duplicates   Unique  \
243  bureau  (1716428, 17)              SK_ID_CURR     1410617   305811   
244  bureau  (1716428, 17)            SK_ID_BUREAU           0  1716428   
245  bureau  (1716428, 17)           CREDIT_ACTIVE     1716424        4   
246  bureau  (1716428, 17)         CREDIT_CURRENCY     1716424        4   
247  bureau  (1716428, 17)             DAYS_CREDIT     1713505     2923   
248  bureau  (1716428, 17)      CREDIT_DAY_OVERDUE     1715486      942   
249  bureau  (1716428, 17)     DAYS_CREDIT_ENDDATE     1702331    14097   
250  bureau  (1716428, 17)       DAYS_ENDDATE_FACT     1713510     2918   
251  bureau  (1716428, 17)  AMT_CREDIT_MAX_OVERDUE     1648176    68252   
252  bureau  (1716428, 17)      CNT_CREDIT_PROLONG     1716418       10   
253  bureau  (1716428, 17)          AMT_CREDIT_SUM     1479719   236709   
254  bureau  (1716428, 17)     AMT_CREDIT_SUM_DEBT     1489890   226538   
255  bureau  (1716428, 17)    AMT_CREDIT_SUM_LIMIT     1664701    51727   
256  bureau  (1716428, 17)  AMT_CREDIT_SUM_OVERDUE     1714812     1616   
257  bureau  (1716428, 17)             CREDIT_TYPE     1716413       15   
258  bureau  (1716428, 17)      DAYS_CREDIT_UPDATE     1713446     2982   
259  bureau  (1716428, 17)             AMT_ANNUITY     1676106    40322   

        Null  Null_ratio     Type  
243        0    0.000000    int64  
244        0    0.000000    int64  
245        0    0.000000   object  
246        0    0.000000   object  
247        0    0.000000    int64  
248        0    0.000000    int64  
249   105553    0.061496  float64  
250   633653    0.369170  float64  
251  1124488    0.655133  float64  
252        0    0.000000    int64  
253       13    0.000008  float64  
254   257669    0.150119  float64  
255   591780    0.344774  float64  
256        0    0.000000  float64  
257        0    0.000000   object  
258        0    0.000000    int64  
259  1226791    0.714735  float64

### 4 ) bureau_balance

Données mensuelles sur les crédits antérieurs de bureau. Chaque ligne correspond à un mois d'un crédit antérieur et un crédit antérieur peut avoir plusieurs lignes de données mensuelles, une pour chaque mois de la durée du crédit. Le jeu bureau_balance comprend la variable SK_ID_BUREAU (clé secondaire) pour faire le lien avec le jeu de données bureau.

In [10]:
## Informations sur le quatrième dataframe: bureau_balance
all_var_info(bureau_balance,'bureau_balance')

Dataset          Shape        Variable  Duplicates  Unique  Null  \
260  bureau_balance  (27299925, 3)    SK_ID_BUREAU    26482530  817395     0   
261  bureau_balance  (27299925, 3)  MONTHS_BALANCE    27299828      97     0   
262  bureau_balance  (27299925, 3)          STATUS    27299917       8     0   

     Null_ratio    Type  
260         0.0   int64  
261         0.0   int64  
262         0.0  object

### 5) previous_application

Précédentes demandes de crédit. Chaque ligne concerne une précédente demande de crédit identifiée par la variable SK_ID_PREV (clé primaire). Chaque crédit en cours (application_train) peut être lié à plusieurs crédits précédents. Le jeu previous_application comprend  la variable SK_ID_CURR (clé secondaire) pour faire le lien avec le jeu de données application_train

In [11]:
## Informations sur le cinquième dataframe: previous_app
all_var_info(previous_app,'previous_application')

Dataset          Shape                     Variable  \
263  previous_application  (1670214, 37)                   SK_ID_PREV   
264  previous_application  (1670214, 37)                   SK_ID_CURR   
265  previous_application  (1670214, 37)           NAME_CONTRACT_TYPE   
266  previous_application  (1670214, 37)                  AMT_ANNUITY   
267  previous_application  (1670214, 37)              AMT_APPLICATION   
268  previous_application  (1670214, 37)                   AMT_CREDIT   
269  previous_application  (1670214, 37)             AMT_DOWN_PAYMENT   
270  previous_application  (1670214, 37)              AMT_GOODS_PRICE   
271  previous_application  (1670214, 37)   WEEKDAY_APPR_PROCESS_START   
272  previous_application  (1670214, 37)      HOUR_APPR_PROCESS_START   
273  previous_application  (1670214, 37)  FLAG_LAST_APPL_PER_CONTRACT   
274  previous_application  (1670214, 37)       NFLAG_LAST_APPL_IN_DAY   
275  previous_application  (1670214, 37)            RATE_DOWN_PAYMENT   
276  previous_application  (1670214, 37)        RATE_INTEREST_PRIMARY   
277  previous_application  (1670214, 37)     RATE_INTEREST_PRIVILEGED   
278  previous_application  (1670214, 37)       NAME_CASH_LOAN_PURPOSE   
279  previous_application  (1670214, 37)         NAME_CONTRACT_STATUS   
280  previous_application  (1670214, 37)                DAYS_DECISION   
281  previous_application  (1670214, 37)            NAME_PAYMENT_TYPE   
282  previous_application  (1670214, 37)           CODE_REJECT_REASON   
283  previous_application  (1670214, 37)              NAME_TYPE_SUITE   
284  previous_application  (1670214, 37)             NAME_CLIENT_TYPE   
285  previous_application  (1670214, 37)          NAME_GOODS_CATEGORY   
286  previous_application  (1670214, 37)               NAME_PORTFOLIO   
287  previous_application  (1670214, 37)            NAME_PRODUCT_TYPE   
288  previous_application  (1670214, 37)                 CHANNEL_TYPE   
289  previous_application  (1670214, 37)             SELLERPLACE_AREA   
290  previous_application  (1670214, 37)         NAME_SELLER_INDUSTRY   
291  previous_application  (1670214, 37)                  CNT_PAYMENT   
292  previous_application  (1670214, 37)             NAME_YIELD_GROUP   
293  previous_application  (1670214, 37)          PRODUCT_COMBINATION   
294  previous_application  (1670214, 37)           DAYS_FIRST_DRAWING   
295  previous_application  (1670214, 37)               DAYS_FIRST_DUE   
296  previous_application  (1670214, 37)    DAYS_LAST_DUE_1ST_VERSION   
297  previous_application  (1670214, 37)                DAYS_LAST_DUE   
298  previous_application  (1670214, 37)             DAYS_TERMINATION   
299  previous_application  (1670214, 37)    NFLAG_INSURED_ON_APPROVAL   

     Duplicates   Unique     Null    Null_ratio     Type  
263           0  1670214        0  0.000000e+00    int64  
264     1331357   338857        0  0.000000e+00    int64  
265     1670210        4        0  0.000000e+00   object  
266     1312254   357960   372235  2.228667e-01  float64  
267     1576329    93885        0  0.000000e+00  float64  
268     1583410    86804        1  5.987257e-07  float64  
269     1640935    29279   895844  5.363648e-01  float64  
270     1576328    93886   385515  2.308177e-01  float64  
271     1670207        7        0  0.000000e+00   object  
272     1670190       24        0  0.000000e+00    int64  
273     1670212        2        0  0.000000e+00   object  
274     1670212        2        0  0.000000e+00    int64  
275     1463180   207034   895844  5.363648e-01  float64  
276     1670065      149  1664263  9.964370e-01  float64  
277     1670188       26  1664263  9.964370e-01  float64  
278     1670189       25        0  0.000000e+00   object  
279     1670210        4        0  0.000000e+00   object  
280     1667292     2922        0  0.000000e+00    int64  
281     1670210        4        0  0.000000e+00   object  
282     1670205        9        0  0.000000e+00   object  
283 

### 6) POS_CASH_BALANCE

Données mensuelles de *prêts en espèces* relatives aux précédentes demandes de crédits. Chaque ligne correspond à une précédente demande de crédit, identifiée par la variable SK_ID_PREV (clé secondaire).

In [12]:
## Informations sur le sixième dataframe: POS_CASH_balance
all_var_info(POS_CASH_balance,'POS_CASH_balance')

Dataset          Shape               Variable  Duplicates  \
300  POS_CASH_balance  (10001358, 8)             SK_ID_PREV     9065033   
301  POS_CASH_balance  (10001358, 8)             SK_ID_CURR     9664106   
302  POS_CASH_balance  (10001358, 8)         MONTHS_BALANCE    10001262   
303  POS_CASH_balance  (10001358, 8)         CNT_INSTALMENT    10001284   
304  POS_CASH_balance  (10001358, 8)  CNT_INSTALMENT_FUTURE    10001278   
305  POS_CASH_balance  (10001358, 8)   NAME_CONTRACT_STATUS    10001349   
306  POS_CASH_balance  (10001358, 8)                 SK_DPD     9997958   
307  POS_CASH_balance  (10001358, 8)             SK_DPD_DEF     9999051   

     Unique   Null  Null_ratio     Type  
300  936325      0    0.000000    int64  
301  337252      0    0.000000    int64  
302      96      0    0.000000    int64  
303      74  26071    0.002607  float64  
304      80  26087    0.002608  float64  
305       9      0    0.000000   object  
306    3400      0    0.000000    int64  
307    2307      0    0.000000    int64

### 7) credit_card_balance

Données mensuelles sur les précédentes *cartes de crédit* liées aux précédentes demandes de crédit. Chaque ligne correspond à un mois de solde de carte de crédit et est liée à une précédente demande de crédit, identifiée par la variable SK_ID_PREV (clé secondaire).

In [13]:
## Informations sur le septième dataframe: credit_card_balance
all_var_info(credit_card_balance,'credit_card_balance')

Dataset          Shape                    Variable  \
308  credit_card_balance  (3840312, 23)                  SK_ID_PREV   
309  credit_card_balance  (3840312, 23)                  SK_ID_CURR   
310  credit_card_balance  (3840312, 23)              MONTHS_BALANCE   
311  credit_card_balance  (3840312, 23)                 AMT_BALANCE   
312  credit_card_balance  (3840312, 23)     AMT_CREDIT_LIMIT_ACTUAL   
313  credit_card_balance  (3840312, 23)    AMT_DRAWINGS_ATM_CURRENT   
314  credit_card_balance  (3840312, 23)        AMT_DRAWINGS_CURRENT   
315  credit_card_balance  (3840312, 23)  AMT_DRAWINGS_OTHER_CURRENT   
316  credit_card_balance  (3840312, 23)    AMT_DRAWINGS_POS_CURRENT   
317  credit_card_balance  (3840312, 23)     AMT_INST_MIN_REGULARITY   
318  credit_card_balance  (3840312, 23)         AMT_PAYMENT_CURRENT   
319  credit_card_balance  (3840312, 23)   AMT_PAYMENT_TOTAL_CURRENT   
320  credit_card_balance  (3840312, 23)    AMT_RECEIVABLE_PRINCIPAL   
321  credit_card_balance  (3840312, 23)               AMT_RECIVABLE   
322  credit_card_balance  (3840312, 23)        AMT_TOTAL_RECEIVABLE   
323  credit_card_balance  (3840312, 23)    CNT_DRAWINGS_ATM_CURRENT   
324  credit_card_balance  (3840312, 23)        CNT_DRAWINGS_CURRENT   
325  credit_card_balance  (3840312, 23)  CNT_DRAWINGS_OTHER_CURRENT   
326  credit_card_balance  (3840312, 23)    CNT_DRAWINGS_POS_CURRENT   
327  credit_card_balance  (3840312, 23)   CNT_INSTALMENT_MATURE_CUM   
328  credit_card_balance  (3840312, 23)        NAME_CONTRACT_STATUS   
329  credit_card_balance  (3840312, 23)                      SK_DPD   
330  credit_card_balance  (3840312, 23)                  SK_DPD_DEF   

     Duplicates   Unique    Null  Null_ratio     Type  
308     3736005   104307       0    0.000000    int64  
309     3736754   103558       0    0.000000    int64  
310     3840216       96       0    0.000000    int64  
311     2492408  1347904       0    0.000000  float64  
312     3840131      181       0    0.000000    int64  
313     3838044     2268  749816    0.195249  float64  
314     3653307   187005       0    0.000000  float64  
315     3838479     1833  749816    0.195249  float64  
316     3671563   168749  749816    0.195249  float64  
317     3528045   312267  305236    0.079482  float64  
318     3677102   163210  767988    0.199981  float64  
319     3657355   182957       0    0.000000  float64  
320     2644473  1195839       0    0.000000  float64  
321     2501434  1338878       0    0.000000  float64  
322     2501304  1339008       0    0.000000  float64  
323     3840267       45  749816    0.195249  float64  
324     3840183      129       0    0.000000    int64  
325     3840300       12  749816    0.195249  float64  
326     3840178      134  749816    0.195249  float64  
327     3840190      122  305236    0.079482  float64  
328     3840305        7       0    0.000000   object  
329     3839395      917       0    0.000000    int64  
330     3839934      378       0    0.000000    int64

### 8)  installments_payments

Historique des paiements pour les précédentes demandes de crédit. Les lignes contiennent les informations relatives aux paiements effectués et aux paiements manqués.

In [14]:
## Informations sur le huitième dataframe: installments_payments
all_var_info(installments_payments,'installments_payments')

Dataset          Shape                Variable  Duplicates  \
331  installments_payments  (13605401, 8)              SK_ID_PREV    12607649   
332  installments_payments  (13605401, 8)              SK_ID_CURR    13265814   
333  installments_payments  (13605401, 8)  NUM_INSTALMENT_VERSION    13605336   
334  installments_payments  (13605401, 8)   NUM_INSTALMENT_NUMBER    13605124   
335  installments_payments  (13605401, 8)         DAYS_INSTALMENT    13602479   
336  installments_payments  (13605401, 8)      DAYS_ENTRY_PAYMENT    13602361   
337  installments_payments  (13605401, 8)          AMT_INSTALMENT    12702862   
338  installments_payments  (13605401, 8)             AMT_PAYMENT    12661165   

     Unique  Null  Null_ratio     Type  
331  997752     0    0.000000    int64  
332  339587     0    0.000000    int64  
333      65     0    0.000000  float64  
334     277     0    0.000000    int64  
335    2922     0    0.000000  float64  
336    3040  2905    0.000214  float64  
337  902539     0    0.000000  float64  
338  944236  2905    0.000214  float64

### 9) sample_submission

In [15]:
## Informations sur le neuvième dataframe: sample_submission
all_var_info(sample_submission,'sample_submission')

Dataset       Shape    Variable  Duplicates  Unique  Null  \
339  sample_submission  (48744, 2)  SK_ID_CURR           0   48744     0   
340  sample_submission  (48744, 2)      TARGET       48743       1     0   

     Null_ratio     Type  
339         0.0    int64  
340         0.0  float64

### 10) HomeCredit_columns_description
Ce dataframe contient la liste des tables et des colonnes avec leurs descriptions.

In [16]:
## Informations sur le dixième dataframe: HomeCredit_columns_description
all_var_info(HomeCredit_columns_description,'HomeCredit_columns_description')

Dataset     Shape     Variable  Duplicates  \
341  HomeCredit_columns_description  (219, 5)   Unnamed: 0           0   
342  HomeCredit_columns_description  (219, 5)        Table         212   
343  HomeCredit_columns_description  (219, 5)          Row          23   
344  HomeCredit_columns_description  (219, 5)  Description          56   
345  HomeCredit_columns_description  (219, 5)      Special         211   

     Unique  Null  Null_ratio    Type  
341     219     0    0.000000   int64  
342       7     0    0.000000  object  
343     196     0    0.000000  object  
344     163     0    0.000000  object  
345       8   133    0.607306  object

In [17]:
HomeCredit_columns_description.head(10)

Unnamed: 0                         Table                 Row  \
0           1  application_{train|test}.csv          SK_ID_CURR   
1           2  application_{train|test}.csv              TARGET   
2           5  application_{train|test}.csv  NAME_CONTRACT_TYPE   
3           6  application_{train|test}.csv         CODE_GENDER   
4           7  application_{train|test}.csv        FLAG_OWN_CAR   
5           8  application_{train|test}.csv     FLAG_OWN_REALTY   
6           9  application_{train|test}.csv        CNT_CHILDREN   
7          10  application_{train|test}.csv    AMT_INCOME_TOTAL   
8          11  application_{train|test}.csv          AMT_CREDIT   
9          12  application_{train|test}.csv         AMT_ANNUITY   

                                         Description Special  
0                           ID of loan in our sample     NaN  
1  Target variable (1 - client with payment diffi...     NaN  
2        Identification if loan is cash or revolving     NaN  
3                               Gender of the client     NaN  
4                      Flag if the client owns a car     NaN  
5                Flag if client owns a house or flat     NaN  
6                  Number of children the client has     NaN  
7                               Income of the client     NaN  
8                          Credit amount of the loan     NaN  
9                                       Loan annuity     NaN

In [18]:
HomeCredit_columns_description["Special"].unique()

array([nan, 'normalized ', 'time only relative to the application',
       'rounded', 'normalized', 'hashed', 'recoded', 'grouped'],
      dtype=object)

Dans un premier temps, on récupère dans une table le nombre de crédits de nos clients ainsi que le montant de ces crédits dans d'autres établissements bancaires à l'aide du dataframe bureau

In [19]:
# Nombre de crédits de chaque client dans d'autres établissements bancaires
previous_credits_count = bureau.groupby('SK_ID_CURR',
                        as_index=False)['SK_ID_BUREAU'].count().rename(columns={'SK_ID_BUREAU': 'Count_previous_credit'})
print(previous_credits_count.shape)
previous_credits_count.head()

(305811, 2)


SK_ID_CURR  Count_previous_credit
0      100001                      7
1      100002                      8
2      100003                      4
3      100004                      2
4      100005                      3

In [20]:
# Somme totale des crédits de chaque client dans d'autres établissements bancaires
previous_credits_amount = bureau.groupby('SK_ID_CURR',
                        as_index=False)['AMT_CREDIT_SUM'].sum().rename(columns={'AMT_CREDIT_SUM': 'Amount_previous_credit'})
print(previous_credits_amount.shape)
previous_credits_amount.head()

(305811, 2)


SK_ID_CURR  Amount_previous_credit
0      100001             1453365.000
1      100002              865055.565
2      100003             1017400.500
3      100004              189037.800
4      100005              657126.000

In [21]:
# Jointure des deux tables précédentes : table Previous_credits
previous_credits= previous_credits_count.merge(previous_credits_amount, on='SK_ID_CURR', how='left')
previous_credits.head(10)

SK_ID_CURR  Count_previous_credit  Amount_previous_credit
0      100001                      7             1453365.000
1      100002                      8              865055.565
2      100003                      4             1017400.500
3      100004                      2              189037.800
4      100005                      3              657126.000
5      100007                      1              146250.000
6      100008                      3              468445.500
7      100009                     18             4800811.500
8      100010                      2              990000.000
9      100011                      4              435228.300

In [22]:
# On ajoute cette nouvelle table au dataframe principal app_train
app_train_bis = app_train.merge(previous_credits, on='SK_ID_CURR', how='left')
app_train_bis.head()


SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   

   AMT_GOODS_PRICE NAME_TYPE_SUITE NAME_INCOME_TYPE  \
0         351000.0   Unaccompanied          Working   
1        1129500.0          Family    State servant   
2         135000.0   Unaccompanied          Working   
3         297000.0   Unaccompanied          Working   
4         513000.0   Unaccompanied          Working   

             NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  \
0  Secondary / secondary special  Single / not married  House / apartment   
1               Higher education               Married  House / apartment   
2  Secondary / secondary special  Single / not married  House / apartment   
3  Secondary / secondary special        Civil marriage  House / apartment   
4  Secondary / secondary special  Single / not married  House / apartment   

   REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  \
0                    0.018801       -9461           -637            -3648.0   
1                    0.003541      -16765          -1188            -1186.0   
2                    0.010032      -19046           -225            -4260.0   
3                    0.008019      -19005          -3039            -9833.0   
4                    0.028663      -19932          -3038            -4311.0   

   DAYS_ID_PUBLISH  OWN_CAR_AGE  FLAG_MOBIL  FLAG_EMP_PHONE  FLAG_WORK_PHONE  \
0            -2120          NaN           1               1                0   
1             -291          NaN           1               1                0   
2            -2531         26.0           1               1                1   
3            -2437          NaN           1               1                0   
4            -3458          NaN           1               1                0   

   ...  FLAG_DOCUMENT_5  FLAG_DOCUMENT_6  FLAG_DOCUMENT_7 FLAG_DOCUMENT_8  \
0  ...                0                0                0               0   
1  ...                0                0                0               0   
2  ...                0                0                0               0   
3  ...                0                0                0               0   
4  ...                0                0                0               1   

   FLAG_DOCUMENT_9  FLAG_DOCUMENT_10  FLAG_DOCUMENT_11 FLAG_DOCUMENT_12  \
0                0                 0                 0                0   
1                0                 0                 0                0   
2                0                 0                 0                0   
3                0                 0                 0                0   
4                0                 0                 0                0   

   FLAG_DOCUMENT_13  FLAG_DOCUMENT_14  FLAG_DOCUMENT_15  FLAG_DOCUMENT_16  \
0                 0                 0                 0                 0   
1                 0                 0                 0                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 0               

In [23]:
app_train_bis["Count_previous_credit"].isna().sum()

44020

Pour 44020 clients, il n y a pas de crédits antérieurs dans d'autres établissements bancaires. Nous allons donc remplacer les NAN par 0 pour le nombre de crédits antérieurs et pour le montant des crédits antérieurs

In [24]:
app_train_bis['Count_previous_credit'] = app_train_bis['Count_previous_credit'].fillna(0)
app_train_bis['Amount_previous_credit'] = app_train_bis['Amount_previous_credit'].fillna(0)
app_train_bis.head()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   

   AMT_GOODS_PRICE NAME_TYPE_SUITE NAME_INCOME_TYPE  \
0         351000.0   Unaccompanied          Working   
1        1129500.0          Family    State servant   
2         135000.0   Unaccompanied          Working   
3         297000.0   Unaccompanied          Working   
4         513000.0   Unaccompanied          Working   

             NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  \
0  Secondary / secondary special  Single / not married  House / apartment   
1               Higher education               Married  House / apartment   
2  Secondary / secondary special  Single / not married  House / apartment   
3  Secondary / secondary special        Civil marriage  House / apartment   
4  Secondary / secondary special  Single / not married  House / apartment   

   REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  \
0                    0.018801       -9461           -637            -3648.0   
1                    0.003541      -16765          -1188            -1186.0   
2                    0.010032      -19046           -225            -4260.0   
3                    0.008019      -19005          -3039            -9833.0   
4                    0.028663      -19932          -3038            -4311.0   

   DAYS_ID_PUBLISH  OWN_CAR_AGE  FLAG_MOBIL  FLAG_EMP_PHONE  FLAG_WORK_PHONE  \
0            -2120          NaN           1               1                0   
1             -291          NaN           1               1                0   
2            -2531         26.0           1               1                1   
3            -2437          NaN           1               1                0   
4            -3458          NaN           1               1                0   

   ...  FLAG_DOCUMENT_5  FLAG_DOCUMENT_6  FLAG_DOCUMENT_7 FLAG_DOCUMENT_8  \
0  ...                0                0                0               0   
1  ...                0                0                0               0   
2  ...                0                0                0               0   
3  ...                0                0                0               0   
4  ...                0                0                0               1   

   FLAG_DOCUMENT_9  FLAG_DOCUMENT_10  FLAG_DOCUMENT_11 FLAG_DOCUMENT_12  \
0                0                 0                 0                0   
1                0                 0                 0                0   
2                0                 0                 0                0   
3                0                 0                 0                0   
4                0                 0                 0                0   

   FLAG_DOCUMENT_13  FLAG_DOCUMENT_14  FLAG_DOCUMENT_15  FLAG_DOCUMENT_16  \
0                 0                 0                 0                 0   
1                 0                 0                 0                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 0               

### Traitement des valeurs manquantes

In [25]:
app_train_bis.isna().sum()

SK_ID_CURR                           0
TARGET                               0
NAME_CONTRACT_TYPE                   0
CODE_GENDER                          0
FLAG_OWN_CAR                         0
FLAG_OWN_REALTY                      0
CNT_CHILDREN                         0
AMT_INCOME_TOTAL                     0
AMT_CREDIT                           0
AMT_ANNUITY                         12
AMT_GOODS_PRICE                    278
NAME_TYPE_SUITE                   1292
NAME_INCOME_TYPE                     0
NAME_EDUCATION_TYPE                  0
NAME_FAMILY_STATUS                   0
NAME_HOUSING_TYPE                    0
REGION_POPULATION_RELATIVE           0
DAYS_BIRTH                           0
DAYS_EMPLOYED                        0
DAYS_REGISTRATION                    0
DAYS_ID_PUBLISH                      0
OWN_CAR_AGE                     202929
FLAG_MOBIL                           0
FLAG_EMP_PHONE                       0
FLAG_WORK_PHONE                      0
FLAG_CONT_MOBILE         

In [26]:
# Pour les colonnes de type "float" et "int", nous utiliserons SimpleImputer pour remplacer les valeurs manquantes par la moyenne
from sklearn.impute import SimpleImputer

col_float = list(app_train_bis.select_dtypes('float64').columns)
col_int = list(app_train_bis.select_dtypes('int64').columns)
imp_mean = SimpleImputer(missing_values=np.nan, strategy='mean')
app_train_bis[col_float] = imp_mean.fit_transform(app_train_bis[col_float])
app_train_bis[col_int] = imp_mean.fit_transform(app_train_bis[col_int])

In [27]:
# Pour les autres colonnes , nous remplacerons les valeurs nulles par "Unknown"
col_object=list(app_train_bis.select_dtypes('object').columns)
imp_object = SimpleImputer(missing_values=np.nan, strategy='constant', fill_value='Unknown')
app_train_bis[col_object] = imp_object.fit_transform(app_train_bis[col_object])

In [28]:
# On verifie qu'il n y a plus de valeur manquante pour app_train_bis
app_train_bis.isna().sum()

SK_ID_CURR                      0
TARGET                          0
NAME_CONTRACT_TYPE              0
CODE_GENDER                     0
FLAG_OWN_CAR                    0
FLAG_OWN_REALTY                 0
CNT_CHILDREN                    0
AMT_INCOME_TOTAL                0
AMT_CREDIT                      0
AMT_ANNUITY                     0
AMT_GOODS_PRICE                 0
NAME_TYPE_SUITE                 0
NAME_INCOME_TYPE                0
NAME_EDUCATION_TYPE             0
NAME_FAMILY_STATUS              0
NAME_HOUSING_TYPE               0
REGION_POPULATION_RELATIVE      0
DAYS_BIRTH                      0
DAYS_EMPLOYED                   0
DAYS_REGISTRATION               0
DAYS_ID_PUBLISH                 0
OWN_CAR_AGE                     0
FLAG_MOBIL                      0
FLAG_EMP_PHONE                  0
FLAG_WORK_PHONE                 0
FLAG_CONT_MOBILE                0
FLAG_PHONE                      0
FLAG_EMAIL                      0
OCCUPATION_TYPE                 0
CNT_FAM_MEMBER

### Analyse des features de app_train_bis

In [29]:
# Nombre de colonnes de chaque type
app_train_bis.dtypes.value_counts()

float64    108
object      16
Name: count, dtype: int64

In [30]:
# Nombre de valeurs uniques dans chaque colonne pour les features categorielles
app_train_bis.select_dtypes('object').apply(pd.Series.nunique, axis = 0)

NAME_CONTRACT_TYPE             2
CODE_GENDER                    3
FLAG_OWN_CAR                   2
FLAG_OWN_REALTY                2
NAME_TYPE_SUITE                8
NAME_INCOME_TYPE               8
NAME_EDUCATION_TYPE            5
NAME_FAMILY_STATUS             6
NAME_HOUSING_TYPE              6
OCCUPATION_TYPE               19
WEEKDAY_APPR_PROCESS_START     7
ORGANIZATION_TYPE             58
FONDKAPREMONT_MODE             5
HOUSETYPE_MODE                 4
WALLSMATERIAL_MODE             8
EMERGENCYSTATE_MODE            3
dtype: int64

### Traitement des variables catégorielles pour app_train_bis

In [31]:
for i in col_object:
    display(i)
    display(app_train_bis[i].unique())

'NAME_CONTRACT_TYPE'

array(['Cash loans', 'Revolving loans'], dtype=object)

'CODE_GENDER'

array(['M', 'F', 'XNA'], dtype=object)

'FLAG_OWN_CAR'

array(['N', 'Y'], dtype=object)

'FLAG_OWN_REALTY'

array(['Y', 'N'], dtype=object)

'NAME_TYPE_SUITE'

array(['Unaccompanied', 'Family', 'Spouse, partner', 'Children',
       'Other_A', 'Unknown', 'Other_B', 'Group of people'], dtype=object)

'NAME_INCOME_TYPE'

array(['Working', 'State servant', 'Commercial associate', 'Pensioner',
       'Unemployed', 'Student', 'Businessman', 'Maternity leave'],
      dtype=object)

'NAME_EDUCATION_TYPE'

array(['Secondary / secondary special', 'Higher education',
       'Incomplete higher', 'Lower secondary', 'Academic degree'],
      dtype=object)

'NAME_FAMILY_STATUS'

array(['Single / not married', 'Married', 'Civil marriage', 'Widow',
       'Separated', 'Unknown'], dtype=object)

'NAME_HOUSING_TYPE'

array(['House / apartment', 'Rented apartment', 'With parents',
       'Municipal apartment', 'Office apartment', 'Co-op apartment'],
      dtype=object)

'OCCUPATION_TYPE'

array(['Laborers', 'Core staff', 'Accountants', 'Managers', 'Unknown',
       'Drivers', 'Sales staff', 'Cleaning staff', 'Cooking staff',
       'Private service staff', 'Medicine staff', 'Security staff',
       'High skill tech staff', 'Waiters/barmen staff',
       'Low-skill Laborers', 'Realty agents', 'Secretaries', 'IT staff',
       'HR staff'], dtype=object)

'WEEKDAY_APPR_PROCESS_START'

array(['WEDNESDAY', 'MONDAY', 'THURSDAY', 'SUNDAY', 'SATURDAY', 'FRIDAY',
       'TUESDAY'], dtype=object)

'ORGANIZATION_TYPE'

array(['Business Entity Type 3', 'School', 'Government', 'Religion',
       'Other', 'XNA', 'Electricity', 'Medicine',
       'Business Entity Type 2', 'Self-employed', 'Transport: type 2',
       'Construction', 'Housing', 'Kindergarten', 'Trade: type 7',
       'Industry: type 11', 'Military', 'Services', 'Security Ministries',
       'Transport: type 4', 'Industry: type 1', 'Emergency', 'Security',
       'Trade: type 2', 'University', 'Transport: type 3', 'Police',
       'Business Entity Type 1', 'Postal', 'Industry: type 4',
       'Agriculture', 'Restaurant', 'Culture', 'Hotel',
       'Industry: type 7', 'Trade: type 3', 'Industry: type 3', 'Bank',
       'Industry: type 9', 'Insurance', 'Trade: type 6',
       'Industry: type 2', 'Transport: type 1', 'Industry: type 12',
       'Mobile', 'Trade: type 1', 'Industry: type 5', 'Industry: type 10',
       'Legal Services', 'Advertising', 'Trade: type 5', 'Cleaning',
       'Industry: type 13', 'Trade: type 4', 'Telecom',
       'I

'FONDKAPREMONT_MODE'

array(['reg oper account', 'Unknown', 'org spec account',
       'reg oper spec account', 'not specified'], dtype=object)

'HOUSETYPE_MODE'

array(['block of flats', 'Unknown', 'terraced house', 'specific housing'],
      dtype=object)

'WALLSMATERIAL_MODE'

array(['Stone, brick', 'Block', 'Unknown', 'Panel', 'Mixed', 'Wooden',
       'Others', 'Monolithic'], dtype=object)

'EMERGENCYSTATE_MODE'

array(['No', 'Unknown', 'Yes'], dtype=object)

In [32]:
# Traitement de NAME_CONTRACT_TYPE
app_train_bis["NAME_CONTRACT_TYPE"].value_counts()

NAME_CONTRACT_TYPE
Cash loans         278232
Revolving loans     29279
Name: count, dtype: int64

In [33]:
# Traitement de NAME_CONTRACT_TYPE
app_train_bis.loc[app_train_bis["NAME_CONTRACT_TYPE"]=="Cash loans","NAME_CONTRACT_TYPE"]=1
app_train_bis.loc[app_train_bis["NAME_CONTRACT_TYPE"]=="Revolving loans","NAME_CONTRACT_TYPE"]=0

In [34]:
# Traitement de CODE_GENDER
app_train_bis["CODE_GENDER"].value_counts()

CODE_GENDER
F      202448
M      105059
XNA         4
Name: count, dtype: int64

In [35]:
# Traitement de CODE_GENDER
# la classe des femmes étant plus nombreuse que celle des hommes , j attribue la valeur femme à la valeur XNA
app_train_bis.loc[app_train_bis["CODE_GENDER"]=="XNA","CODE_GENDER"]="F"
# Puis j'attribue la valeur 1 à femme et la valeur 0 à homme
app_train_bis.loc[app_train_bis["CODE_GENDER"]=="F","CODE_GENDER"]=1
app_train_bis.loc[app_train_bis["CODE_GENDER"]=="M","CODE_GENDER"]=0
app_train_bis["CODE_GENDER"].unique()


array([0, 1], dtype=object)

In [36]:
# Traitement de FLAG_OWN_CAR ( possède une voiture) et FLAG_OWN_REALTY (propriétaire d'une maison ou appart): 1 pour non et 0 pour oui
app_train_bis.loc[app_train_bis["FLAG_OWN_CAR"]=="N","FLAG_OWN_CAR"]=1
app_train_bis.loc[app_train_bis["FLAG_OWN_CAR"]=="Y","FLAG_OWN_CAR"]=0
app_train_bis.loc[app_train_bis["FLAG_OWN_REALTY"]=="N","FLAG_OWN_REALTY"]=1
app_train_bis.loc[app_train_bis["FLAG_OWN_REALTY"]=="Y","FLAG_OWN_REALTY"]=0

In [37]:
# Traitement de NAME_TYPE_SUITE ( qui accompagnait le client lorsqu'il a demandé le prêt): regroupement
app_train_bis.loc[(app_train_bis["NAME_TYPE_SUITE"]=="Spouse, partner")|
                  (app_train_bis["NAME_TYPE_SUITE"]=="Children"),"NAME_TYPE_SUITE"]="Family"
app_train_bis.loc[(app_train_bis["NAME_TYPE_SUITE"]=="Other_A")|
                  (app_train_bis["NAME_TYPE_SUITE"]=="Other_B")|
                  (app_train_bis["NAME_TYPE_SUITE"]=="Group of people"),"NAME_TYPE_SUITE"]="Other"
app_train_bis["NAME_TYPE_SUITE"].unique()

array(['Unaccompanied', 'Family', 'Other', 'Unknown'], dtype=object)

In [38]:
app_train_bis["NAME_TYPE_SUITE"].value_counts(True)

NAME_TYPE_SUITE
Unaccompanied    0.808186
Family           0.178159
Other            0.009453
Unknown          0.004201
Name: proportion, dtype: float64

In [39]:
app_train_bis.loc[(app_train_bis["NAME_TYPE_SUITE"]=="Family")|
                  (app_train_bis["NAME_TYPE_SUITE"]=="Other")|
                  (app_train_bis["NAME_TYPE_SUITE"]=="Unknown"),"NAME_TYPE_SUITE"]="Accompanied"

In [40]:
# On met 0 pour les personnes venues seules et 1 pour les personnes accompagnées
app_train_bis.loc[app_train_bis["NAME_TYPE_SUITE"]=="Unaccompanied","NAME_TYPE_SUITE"]=0
app_train_bis.loc[app_train_bis["NAME_TYPE_SUITE"]=="Accompanied","NAME_TYPE_SUITE"]=1
app_train_bis["NAME_TYPE_SUITE"].value_counts(True)


NAME_TYPE_SUITE
0    0.808186
1    0.191814
Name: proportion, dtype: float64

In [41]:
# Traitement de NAME_INCOME_TYPE ( type de revenus): regroupement*
app_train_bis.loc[(app_train_bis["NAME_INCOME_TYPE"]=="State servant")|
                  (app_train_bis["NAME_INCOME_TYPE"]=="Commercial associate")|
                  (app_train_bis["NAME_INCOME_TYPE"]=="Businessman"),"NAME_INCOME_TYPE"]="Working"
app_train_bis["NAME_INCOME_TYPE"].unique()

array(['Working', 'Pensioner', 'Unemployed', 'Student', 'Maternity leave'],
      dtype=object)

In [42]:
app_train_bis["NAME_INCOME_TYPE"].value_counts(True)

NAME_INCOME_TYPE
Working            0.819821
Pensioner          0.180033
Unemployed         0.000072
Student            0.000059
Maternity leave    0.000016
Name: proportion, dtype: float64

In [43]:
app_train_bis.loc[(app_train_bis["NAME_INCOME_TYPE"]=="Pensioner")|
                  (app_train_bis["NAME_INCOME_TYPE"]=="Unemployed")|
                  (app_train_bis["NAME_INCOME_TYPE"]=="Student")|
                  (app_train_bis["NAME_INCOME_TYPE"]=="Maternity leave"),"NAME_INCOME_TYPE"]="Not Working"
app_train_bis["NAME_INCOME_TYPE"].value_counts(True)


NAME_INCOME_TYPE
Working        0.819821
Not Working    0.180179
Name: proportion, dtype: float64

In [44]:
# On met 0 pour les personnes qui travaillent et 1 pour les personnes qui ne travaillent pas
app_train_bis.loc[app_train_bis["NAME_INCOME_TYPE"]=="Working","NAME_INCOME_TYPE"]=0
app_train_bis.loc[app_train_bis["NAME_INCOME_TYPE"]=="Not Working","NAME_INCOME_TYPE"]=1
app_train_bis["NAME_INCOME_TYPE"].value_counts(True)

NAME_INCOME_TYPE
0    0.819821
1    0.180179
Name: proportion, dtype: float64

In [45]:
# Traitement de NAME_EDUCATION_TYPE ( education) : regroupement
app_train_bis.loc[(app_train_bis["NAME_EDUCATION_TYPE"]=="Secondary / secondary special")|
                  (app_train_bis["NAME_EDUCATION_TYPE"]=="Lower secondary")|
                  (app_train_bis["NAME_EDUCATION_TYPE"]=="Incomplete higher"),"NAME_EDUCATION_TYPE"]="Secondary"
app_train_bis.loc[(app_train_bis["NAME_EDUCATION_TYPE"]=="Academic degree"),"NAME_EDUCATION_TYPE"]="Higher education"
app_train_bis["NAME_EDUCATION_TYPE"].unique()

array(['Secondary', 'Higher education'], dtype=object)

In [46]:
# On met 1 pour les personnes qui ont fait des études secondaires et 0 pour les personnes qui ont fait des études supérieures
app_train_bis.loc[app_train_bis["NAME_EDUCATION_TYPE"]=="Secondary","NAME_EDUCATION_TYPE"]=1
app_train_bis.loc[app_train_bis["NAME_EDUCATION_TYPE"]=="Higher education","NAME_EDUCATION_TYPE"]=0
app_train_bis["NAME_EDUCATION_TYPE"].value_counts(True)

NAME_EDUCATION_TYPE
1    0.756018
0    0.243982
Name: proportion, dtype: float64

In [47]:
# Traitement de NAME_FAMILY_STATUS (situation familiale)
app_train_bis.loc[(app_train_bis["NAME_FAMILY_STATUS"]=="Single / not married")|
                  (app_train_bis["NAME_FAMILY_STATUS"]=="Widow")|
                  (app_train_bis["NAME_FAMILY_STATUS"]=="Separated")|
                  (app_train_bis["NAME_FAMILY_STATUS"]=="Unknown"),"NAME_FAMILY_STATUS"]="Single"
app_train_bis.loc[app_train_bis["NAME_FAMILY_STATUS"]=="Civil marriage","NAME_FAMILY_STATUS"]="Married"

app_train_bis["NAME_FAMILY_STATUS"].unique()

array(['Single', 'Married'], dtype=object)

In [48]:
# On met 1 pour les personnes mariées et 0 pour les personnes célibataires
app_train_bis.loc[app_train_bis["NAME_FAMILY_STATUS"]=="Married","NAME_FAMILY_STATUS"]=1
app_train_bis.loc[app_train_bis["NAME_FAMILY_STATUS"]=="Single","NAME_FAMILY_STATUS"]=0
app_train_bis["NAME_FAMILY_STATUS"].value_counts(True)

NAME_FAMILY_STATUS
1    0.735606
0    0.264394
Name: proportion, dtype: float64

In [49]:
# Traitement de NAME_HOUSING_TYPE (type de logement)
app_train_bis.loc[(app_train_bis["NAME_HOUSING_TYPE"]=="Municipal apartment")|
                  (app_train_bis["NAME_HOUSING_TYPE"]=="Office apartment")|
                  (app_train_bis["NAME_HOUSING_TYPE"]=="Co-op apartment"),"NAME_HOUSING_TYPE"]="Rented apartment"
app_train_bis["NAME_HOUSING_TYPE"].unique()


array(['House / apartment', 'Rented apartment', 'With parents'],
      dtype=object)

In [50]:
app_train_bis["NAME_HOUSING_TYPE"].value_counts(True)

NAME_HOUSING_TYPE
House / apartment    0.887344
Rented apartment     0.064398
With parents         0.048258
Name: proportion, dtype: float64

In [51]:
# Traitement de OCCUPATION_TYPE (type de travail) : regroupement
app_train_bis.loc[(app_train_bis["OCCUPATION_TYPE"]=="Laborers")|
                  (app_train_bis["OCCUPATION_TYPE"]=="Core staff")|
                  (app_train_bis["OCCUPATION_TYPE"]=="Drivers")|
                  (app_train_bis["OCCUPATION_TYPE"]=="Sales staff")|
                  (app_train_bis["OCCUPATION_TYPE"]=="Cleaning staff")|
                  (app_train_bis["OCCUPATION_TYPE"]=="Cooking staff")|
                  (app_train_bis["OCCUPATION_TYPE"]=="Private service staff")|
                  (app_train_bis["OCCUPATION_TYPE"]=="Security staff")|
                  (app_train_bis["OCCUPATION_TYPE"]=="Waiters/barmen staff")|
                  (app_train_bis["OCCUPATION_TYPE"]=="Low-skill Laborers")|
                  (app_train_bis["OCCUPATION_TYPE"]=="Secretaries"),"OCCUPATION_TYPE"]="Job low salary"
app_train_bis.loc[(app_train_bis["OCCUPATION_TYPE"]=="Accountants")|
                  (app_train_bis["OCCUPATION_TYPE"]=="Managers")|
                  (app_train_bis["OCCUPATION_TYPE"]=="Medicine staff")|
                  (app_train_bis["OCCUPATION_TYPE"]=="High skill tech staff")|
                  (app_train_bis["OCCUPATION_TYPE"]=="Realty agents")|
                  (app_train_bis["OCCUPATION_TYPE"]=="IT staff")|
                  (app_train_bis["OCCUPATION_TYPE"]=="HR staff"),"OCCUPATION_TYPE"]="Job high salary"

app_train_bis["OCCUPATION_TYPE"].value_counts(True)

OCCUPATION_TYPE
Job low salary     0.514385
Unknown            0.313455
Job high salary    0.172160
Name: proportion, dtype: float64

In [52]:
# Traitement de ORGANIZATION_TYPE (type d'organisation où travaille le client): regroupement
app_train_bis.loc[(app_train_bis["ORGANIZATION_TYPE"]=="Industry: type 1")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Industry: type 2")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Industry: type 3")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Industry: type 4")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Industry: type 5")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Industry: type 6")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Industry: type 7")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Industry: type 8")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Industry: type 9")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Industry: type 10")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Industry: type 11")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Industry: type 12")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Industry: type 13"),"ORGANIZATION_TYPE"]="Industry"
app_train_bis.loc[(app_train_bis["ORGANIZATION_TYPE"]=="Trade: type 1")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Trade: type 2")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Trade: type 3")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Trade: type 4")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Trade: type 5")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Trade: type 6")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Trade: type 7"),"ORGANIZATION_TYPE"]="Trade"
app_train_bis.loc[(app_train_bis["ORGANIZATION_TYPE"]=="Transport: type 1")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Transport: type 2")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Transport: type 3")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Transport: type 4"),"ORGANIZATION_TYPE"]="Transport"
app_train_bis.loc[(app_train_bis["ORGANIZATION_TYPE"]=="Business Entity Type 1")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Business Entity Type 2")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Business Entity Type 3"),"ORGANIZATION_TYPE"]="Business"
app_train_bis.loc[(app_train_bis["ORGANIZATION_TYPE"]=="School")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Government")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Kindergarten")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Military")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Services")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Security Ministries")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="University")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Police")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Postal")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Emergency")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Electricity"),"ORGANIZATION_TYPE"]="Public services"
app_train_bis.loc[(app_train_bis["ORGANIZATION_TYPE"]=="Religion")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="XNA")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Medicine")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Self-employed")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Construction")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Housing")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Security")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Agriculture")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Restaurant")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Culture")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Hotel")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Bank")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Insurance")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Mobile")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Legal Services")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Advertising")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Cleaning")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Telecom")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Realtor"),"ORGANIZATION_TYPE"]="Other"


app_train_bis["ORGANIZATION_TYPE"].value_counts(True)


ORGANIZATION_TYPE
Other              0.473710
Business           0.274881
Public services    0.129085
Trade              0.046551
Industry           0.046538
Transport          0.029235
Name: proportion, dtype: float64

In [53]:
app_train_bis.loc[(app_train_bis["ORGANIZATION_TYPE"]=="Trade")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Industry")|
                  (app_train_bis["ORGANIZATION_TYPE"]=="Transport"),"ORGANIZATION_TYPE"]="Trade_Industry_Transport"
app_train_bis["ORGANIZATION_TYPE"].value_counts(True)

ORGANIZATION_TYPE
Other                       0.473710
Business                    0.274881
Public services             0.129085
Trade_Industry_Transport    0.122324
Name: proportion, dtype: float64

In [54]:
# FONDKAPREMONT_MODE,HOUSETYPE_MODE,WALLSMATERIAL_MODE,EMERGENCYSTATE_MODE sont des informations normalisées sur le logement dans
# lequel vit le client, je décide de les supprimer
display(app_train_bis.shape)
app_train_bis.drop(columns=["FONDKAPREMONT_MODE","HOUSETYPE_MODE","WALLSMATERIAL_MODE","EMERGENCYSTATE_MODE"],inplace=True)
display(app_train_bis.shape)

(307511, 124)

(307511, 120)

In [55]:
# Nombre de valeurs uniques dans chaque colonne pour les features categorielles après nettoyage
app_train_bis.select_dtypes('object').apply(pd.Series.nunique, axis = 0)

NAME_CONTRACT_TYPE            2
CODE_GENDER                   2
FLAG_OWN_CAR                  2
FLAG_OWN_REALTY               2
NAME_TYPE_SUITE               2
NAME_INCOME_TYPE              2
NAME_EDUCATION_TYPE           2
NAME_FAMILY_STATUS            2
NAME_HOUSING_TYPE             3
OCCUPATION_TYPE               3
WEEKDAY_APPR_PROCESS_START    7
ORGANIZATION_TYPE             4
dtype: int64

In [56]:
app_train_bis.head()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0    100002.0     1.0                  1           0            1   
1    100003.0     0.0                  1           1            1   
2    100004.0     0.0                  0           0            0   
3    100006.0     0.0                  1           1            1   
4    100007.0     0.0                  1           0            1   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               0           0.0          202500.0    406597.5      24700.5   
1               1           0.0          270000.0   1293502.5      35698.5   
2               0           0.0           67500.0    135000.0       6750.0   
3               0           0.0          135000.0    312682.5      29686.5   
4               0           0.0          121500.0    513000.0      21865.5   

   AMT_GOODS_PRICE NAME_TYPE_SUITE NAME_INCOME_TYPE NAME_EDUCATION_TYPE  \
0         351000.0               0                0                   1   
1        1129500.0               1                0                   0   
2         135000.0               0                0                   1   
3         297000.0               0                0                   1   
4         513000.0               0                0                   1   

  NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  REGION_POPULATION_RELATIVE  \
0                  0  House / apartment                    0.018801   
1                  1  House / apartment                    0.003541   
2                  0  House / apartment                    0.010032   
3                  1  House / apartment                    0.008019   
4                  0  House / apartment                    0.028663   

   DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  DAYS_ID_PUBLISH  OWN_CAR_AGE  \
0     -9461.0         -637.0            -3648.0          -2120.0    12.061091   
1    -16765.0        -1188.0            -1186.0           -291.0    12.061091   
2    -19046.0         -225.0            -4260.0          -2531.0    26.000000   
3    -19005.0        -3039.0            -9833.0          -2437.0    12.061091   
4    -19932.0        -3038.0            -4311.0          -3458.0    12.061091   

   FLAG_MOBIL  FLAG_EMP_PHONE  FLAG_WORK_PHONE  ...  FLAG_DOCUMENT_5  \
0         1.0             1.0              0.0  ...              0.0   
1         1.0             1.0              0.0  ...              0.0   
2         1.0             1.0              1.0  ...              0.0   
3         1.0             1.0              0.0  ...              0.0   
4         1.0             1.0              0.0  ...              0.0   

   FLAG_DOCUMENT_6  FLAG_DOCUMENT_7 FLAG_DOCUMENT_8  FLAG_DOCUMENT_9  \
0              0.0              0.0             0.0              0.0   
1              0.0              0.0             0.0              0.0   
2              0.0              0.0             0.0              0.0   
3              0.0              0.0             0.0              0.0   
4              0.0              0.0             1.0              0.0   

   FLAG_DOCUMENT_10  FLAG_DOCUMENT_11 FLAG_DOCUMENT_12  FLAG_DOCUMENT_13  \
0               0.0               0.0              0.0               0.0   
1               0.0               0.0              0.0               0.0   
2               0.0               0.0              0.0               0.0   
3               0.0               0.0              0.0               0.0   
4               0.0               0.0              0.0               0.0   

   FLAG_DOCUMENT_14  FLAG_DOCUMENT_15  FLAG_DOCUMENT_16  FLAG_DOCUMENT_17  \
0               0.0               0.0               0.0               0.0   
1               0.0               0.0               0.0               0.0   
2               0.0               0.0               0.0               0.0   
3               0.0               0.0               0.0               0.0   
4               0.0               0.0         

In [57]:
app_train_bis.to_csv("app_train_bis.csv", index=False)